### ISSUE TITLE
Albedo (and Normal) AOV support for improved post denoising

### ISSUE THREAD

**Issue by jackjt8** [llbit/chunky#536]
[Feature request]

While many denoisers can operate without **albedo and normal AOVs**, the additional information that these AOVs can provide to a denoiser can make significant improvements to the final image by helping **preserve fine details** and **textures**. Of these two, the albedo image usually provides the biggest quality improvement.

[Albedo and normal AOVs descriptions](https://openimagedenoise.github.io/documentation.html)

While support for these AOVs would not alone improve render times or quality, it would enable users to make better use of desnoisers like [Nvidia's AI Desnoiser (DeclanRussell)](https://github.com/DeclanRussell/NvidiaAIDenoiser) or [Intel Open Image Denoise](https://openimagedenoise.github.io/), though the latter supports all CPUs with SSE4.1+ support and could(?) be implemented into Chunky.

There is also a discussion over at the LuxCoreRender Forums regarding Intel OIDN and the AOVs which might be useful: [Here](http://forums.luxcorerender.org/viewtopic.php?f=5&t=872)

**Comment by leMaik** [comment]
I played around with this a little (not sure if I got the albedo or normal renderer right, though). These are my results:

|Start Image|Denoise|Denoise+Albedo|Denoise+Albedo+Normals|
|---|---|---|---|
|![default-10](https://user-images.githubusercontent.com/5544859/62430408-b7ac4b00-b71c-11e9-8b93-c31a7b21143d.png) (100 SPP)|![only-image](https://user-images.githubusercontent.com/5544859/62430412-bf6bef80-b71c-11e9-9273-6ca4c6f04f7d.png)|![with-albedo](https://user-images.githubusercontent.com/5544859/62430415-c561d080-b71c-11e9-82f9-ca8359a718f6.png)|![with-everything](https://user-images.githubusercontent.com/5544859/62430417-c8f55780-b71c-11e9-8978-28762da5943f.png)
|![default-1000](https://user-images.githubusercontent.com/5544859/62430427-e32f3580-b71c-11e9-8268-9a15f10c1762.png) (1000 SPP)|||![test](https://user-images.githubusercontent.com/5544859/62430434-ede9ca80-b71c-11e9-8efe-1bdb967958c2.png)|

I'm not sure if I got the Albedo renderer and the normal renderer right, though. The normal renderer doesn't change much in the images above.

Normal map:
![normal](https://user-images.githubusercontent.com/5544859/62430445-09ed6c00-b71d-11e9-884d-6e199d1c393f.png)

Albedo image (I used the preview renderer with flat shading disabled, so no shadows):
![albedo](https://user-images.githubusercontent.com/5544859/62430464-30130c00-b71d-11e9-91e7-5d3b5a53ee1c.png)

**Comment by jackjt8** [comment]
Generally speaking I wouldn't use the denoisers for scenes lit directly by the sun and sky as these tend to be fine at low SPP. Instead the AI denosiers tend to be most beneficial when dealing with emitters or indirect lighting conditions where noise is much more common. I'm currently testing with a village (Minecraft 1.12 seed WASWAS gives me a village near spawn).

---

Now, I'm no expert but the normal map seems to be ok, and given the Intel Open Image Denoiser Example, the water should be ok as it is represented correctly as a surface. One part stands out to me however 'the normal image should be anti-aliased', given how your Denoise+Albedo+Normals result is a bit blurier than the Denoise+Albedo result it leads me to believe that the normal might just need to be anti-aliased if not already done. Otherwise there might be some other issue with it.

The Albedo is the tricky one and probably the most useful if we manage to get it right... I believe you are on the right track but I'm seeing mixed results, both in your provided examples and with mine. In some areas where the denoisers blurs the image a lot (mainly distant objects) the Albedo appears to worsen it, where as in some other areas it does seem to bring some detail back.

Now I wasn't able to figure out how to disable the flat shading and disable the crosshair but otherwise it seems to work fine and it's producing similar results/artifacts to what you got.

Albedo
![DN_test_albedo-0](https://user-images.githubusercontent.com/3740917/62478818-ed484700-b7a3-11e9-8191-800f6bbc6932.png)


| SPP | Base | Denoise | +Albedo |
| ------------- | ------------- | ------------- | ------------- |
| 10 | ![DN_test-10](https://user-images.githubusercontent.com/3740917/62478867-136de700-b7a4-11e9-8eb8-78f12948df91.png) |  ![DN_test-10_dn](https://user-images.githubusercontent.com/3740917/62478883-208ad600-b7a4-11e9-9139-20a252f7b901.png) | ![DN_test-10_dn_a](https://user-images.githubusercontent.com/3740917/62478887-241e5d00-b7a4-11e9-8089-f720d8563f23.png) |
| 1000 | ![DN_test-1000](https://user-images.githubusercontent.com/3740917/62479000-6778cb80-b7a4-11e9-9a72-4652da107904.png) | ![DN_test-1000_dn](https://user-images.githubusercontent.com/3740917/62479009-6ba4e900-b7a4-11e9-9356-12ea30def57c.png) | ![DN_test-1000_dn_a](https://user-images.githubusercontent.com/3740917/62479017-6fd10680-b7a4-11e9-8d82-4f0834938957.png) |
| 10000 | ![DN_test-10000](https://user-images.githubusercontent.com/3740917/62479698-f2a69100-b7a5-11e9-9b4f-2dc99366077f.png) | ![DN_test-10000_dn](https://user-images.githubusercontent.com/3740917/62479707-f803db80-b7a5-11e9-8b6e-792477634760.png) | ![DN_test-10000_dn_a](https://user-images.githubusercontent.com/3740917/62479724-fdf9bc80-b7a5-11e9-8941-783eee3d1326.png) |

---

Thanks for your work on this, it's really helpful.

**Comment by leMaik** [comment]
@jackjt8 Nice to hear from you! :tada: Maybe we can chat about this sometime in the Chunky Discord. I'll push my code soon as a Chunky plugin. It just needs some cleanup (currently, I have one plugin for each step – we might be able to put that into one plugin and maybe even call the denoiser from there).

Indeed, the albedo map brings _some_ detail (e.g. dirt and foreground leaves in my examples). Maybe the other leaves are blurry because they are pretty noisy even if they are fully rendered? :thinking: Same as the grass texture?

**Comment by jackjt8** [comment]
Certainly, I'm really active on the discord and your efforts have been a huge help. Hopefully with some more work from both of us we can really nail this.

Merging the plugins into one would certainly save some time. I'm not sure about automatically calling the Denoiser given the mixed results and the system resources spike (OIDN especially eats system RAM) but it might be nice to have a checkbox for auto-denoising if possible otherwise just a button to export the Albedo and Normal.

Yeah, that might just be the issue given what I'm seeing in my tests too however it's hard to tell at such a low resolution. I tend to supersample all my renders: say if I'm targetting 1080p 1024SPP I would instead render at 2160p 256SPP. They tend to take the same amount of time to render, and look similar, but the additional resolution really helps with denoising, both through using older techniques like Gaussian, median, etc. blurs and AI denoising. Might be worth using this mindset for our testing moving forward.
